### ISSUE TITLE
Restore allocate_varlen_buffer 

### ISSUE THREAD

**Issue by pearu** [xnd-project/rbc#87]
The usage of `allocate_varlen_buffer` was replaced with `calloc` in PR #84 as a temporary solution to the issue #75. However, not being able to use functions defined by Omniscidb server blocks several planned features: cuda enabled UDFs, calling library functions (that are linked to the server) from UDFs, etc. So, we'll need to restore the possibility to use `allocate_varlen_buffer` or any other low-level function from rbc generated LLVM IR strings.

A few items to investigate:
1) Do the order of declared functions matter on LLVM IR?
2) Is the function `allocate_varlen_buffer` being declared more than once?
3) Find the usage of `locals` on `compiler.compile_extra`

_Originally posted by @guilhermeleobas in https://github.com/xnd-project/rbc/issues/75#issuecomment-635535014 and edited by @pearu_

**Comment by guilhermeleobas** [comment]
Steps to reproduce:

1. Apply this diff
```diff
diff --git a/rbc/omnisci_array.py b/rbc/omnisci_array.py
index 7ea7c66..cb121f0 100644
--- a/rbc/omnisci_array.py
+++ b/rbc/omnisci_array.py
@@ -69,7 +69,7 @@ def omnisci_array_constructor(context, builder, sig, args):
     alloc_fnty = ir.FunctionType(int8_t.as_pointer(), [int64_t, int64_t])
     # see https://github.com/xnd-project/rbc/issues/75
     alloc_fn = builder.module.get_or_insert_function(
-        alloc_fnty, name="calloc")
+        alloc_fnty, name="allocate_varlen_buffer")
     ptr8 = builder.call(alloc_fn, [element_count, element_size])
     builder_buffers[builder].append(ptr8)
     ptr = builder.bitcast(ptr8, context.get_value_type(ptr_type))
```

2. Run the following test:
```
pytest -sv rbc/tests/test_omnisci_array.py::test_np_cumsum -x --capture=no --disable-warnings
```

**Comment by guilhermeleobas** [comment]
I was able to reproduce this error on Numba. I think it will be easier to solve the problem from this example:
https://gist.github.com/guilhermeleobas/a5e7204f325b526bfe7f01895e0cbc68

**Comment by guilhermeleobas** [comment]
xref https://github.com/numba/numba/issues/6012

**Comment by guilhermeleobas** [comment]
I was finally able to compile a code using `allocate_varlen_buffer`. For that, I had to change a few internals:

RBC:
- `omnisci_array.py::omnisci_array_constructor`: change `calloc` by `allocate_varlen_buffer`:

```python
alloc_fn = builder.module.get_or_insert_function(
    alloc_fnty, name="allocate_varlen_buffer")
alloc_fn.linkage = 'external'
alloc_fn.attributes.add('noinline')
alloc_fn.attributes.add('optnone')
```

Numba:
- `codegen.py::JITCPUCodegen::set_env` must return None rather than do anything else.
- `codegen.py::JITCodeLibrary::get_pointer_to_function` must return 0 (do nothing)
- `codegen.py::JITCodeLibrary::_finalize_specific`: comment call to `self._codegen._engine.finalize_object()`

Numba diff:
```python
diff --git a/numba/core/codegen.py b/numba/core/codegen.py
index 1e05c12ba..ddf5525cb 100644
--- a/numba/core/codegen.py
+++ b/numba/core/codegen.py
@@ -503,16 +503,10 @@ class JITCodeLibrary(CodeLibrary):
               library.
             - non-zero if the symbol is defined.
         """
-        self._ensure_finalized()
-        ee = self._codegen._engine
-        if not ee.is_symbol_defined(name):
-            return 0
-        else:
-            return self._codegen._engine.get_function_address(name)
+        return 0

     def _finalize_specific(self):
         self._codegen._scan_and_fix_unresolved_refs(self._final_module)
-        self._codegen._engine.finalize_object()


 class RuntimeLinker(object):
@@ -863,10 +857,7 @@ class JITCPUCodegen(BaseCPUCodegen):

         Update the GlobalVariable named *env_name* to the address of *env*.
         """
-        gvaddr = self._engine.get_global_value_address(env_name)
-        envptr = (ctypes.c_void_p * 1).from_address(gvaddr)
-        envptr[0] = ctypes.c_void_p(id(env))
-
+        return None

 def initialize_llvm():
     """Safe to use multiple times.
```

A few items to investigate:
- Compilation seems to break because `NoPythonBackend` calls `NativeLowering`. Do we need `NativeLowering`?
- Other targets (CUDA and ROC) don't seem to suffer from this problem. Should we create our own compiler?
- In the end, what I did was to prevent Numba from calling LLVM functions which internally verifies if the module is final (i.e. ready to be executed), which is not always true our case.
- Changing `allocate_varlen_buffer` linkage to `external` didn't seem to suffer any effect.

**Comment by pearu** [comment]
@guilhermeleobas , here's an example how to overwrite `A` method of a `B` instance that has a method calling`A` instance method:
```
class A:

    def mth(self):
        return 1

class B:

    _A_cls = A

    def mth(self):
        a = self._A_cls()
        return a.mth()

b = B()

class MyA:

    def mth(self):
        return 2

class Hack:

    def __init__(self, b, A=A):
        self.A = A
        self.b_mth = b._A_cls.__dict__['mth']
        self.b_mth_code_orig = self.b_mth.__code__
        
    def __enter__(self):
        self.b_mth.__code__ = self.A.mth.__code__

    def __exit__(self, *args):
        self.b_mth.__code__ = self.b_mth_code_orig


assert b.mth() == 1

with Hack(b, MyA):
    assert b.mth() == 2

assert b.mth() == 1
```

**Comment by guilhermeleobas** [comment]
Closed by #236
